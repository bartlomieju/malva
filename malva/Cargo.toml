[package]
name = "malva"
version = "0.0.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "CSS/SCSS/Sass/Less formatter."
repository = "https://github.com/g-plane/malva"
license = "MIT"

[dependencies]
itertools = "0.11"
raffia = "0.4"
tiny_pretty = { path = "../../tiny_pretty" }
