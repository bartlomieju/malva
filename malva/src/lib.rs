mod ctx;
mod doc_gen;
pub mod options;
