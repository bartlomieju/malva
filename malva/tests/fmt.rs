use insta::{assert_snapshot, glob, Settings};
use malva::{config::FormatOptions, format_text};
use raffia::{ast::Stylesheet, Parser, ParserBuilder, Syntax};
use std::fs;

#[test]
fn fmt_snapshot() {
    glob!("fmt/**/*.{css,scss,sass,less}", |path| {
        let input = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };

        let mut comments = vec![];
        let mut parser = ParserBuilder::new(&input)
            .syntax(syntax)
            .comments(&mut comments)
            .build();
        parser
            .parse::<Stylesheet>()
            .map_err(|error| {
                format!(
                    "failed to parse '{}': {} from {} to {}",
                    path.display(),
                    error.kind,
                    error.span.start,
                    error.span.end
                )
            })
            .unwrap();
        let options = if let Some(config) = comments
            .first()
            .and_then(|comment| comment.content.strip_prefix("cfg"))
        {
            toml::from_str::<FormatOptions>(config).unwrap()
        } else {
            Default::default()
        };

        let output = format_text(&input, syntax, &options).unwrap();
        assert!(
            !output.contains(" \n"),
            "'{}' has trailing whitespace",
            path.display()
        );
        let mut regression_parser = Parser::new(&output, syntax);
        assert!(
            regression_parser.parse::<Stylesheet>().is_ok(),
            "'{}' has syntax errors after formatted",
            path.display()
        );

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_snapshot!(name, output);
        });
    });
}
