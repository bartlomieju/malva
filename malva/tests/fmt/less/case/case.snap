---
source: malva/tests/fmt.rs
---
// Convention in this test file:
// - The case should be preserved for things prefixed with "Keep".
// - The case should always be preserved for element names and attribute names
//   in selectors, as well as function names and property keywords.
// - Other things should mostly be lowercase.

html#KeepId.KeepClass,
a[HREF=KeepAttrValue]:hover::first-letter,
svg[viewBox] lineargradient,
:not(:nth-child(2n+1)) {
  color: #aabbcc;
  background-image: url("KeepString");
  margin: 5px 0.2e10em;
  --Keep-custom-Prop: red;
  background: Var(--Keep-custom-Prop);
  animation-name: KeepAnimationName;
  important: something !important;
  font-family: INITIAL;
  padding: UNSET;
  border: INHERIT;
  all: REVERT;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

@keyframes KeepAnimationName {
  from {
    prop: val;
  }

  @{KeepInterpolationVar} {
    prop: val;
  }

  to {
    prop: val;
  }
}

@custom-media --KeepCustomMedia screen and (width >= 768px);
@media (--KeepCustomMedia) {}

@KeepDetachedRuleset: /*:*/ {
  BACKGROUND: RED;
};

@KeepTopLevelVar: val;

.Keep(@Keep: 12e03px) when (@Keep = Case) /*:*/ {
  @KeepVar: KeepName; /*:*/
  @{KeepInterpolationVar}: val;

  @{Keep}[@{Keep}][@{Keep}^=@{Keep-1A}]:@{Keep} {
    prop: val;
  }

  &Keep & element {
    prop: val;
  }

  @KeepDetachedRuleset();

  &:extend(.Keep all);
  &:extend(.Keep all);

  .Keep;
  .Keep();
  .Keep(4px) !important;
  .Keep() when (@Keep = Keep) {}
  .Keep() when (@Keep = 12px) {}
  .Keep() when (@Keep = "12px") {}
  .Keep() when (@Keep = "12PX") {}
  .Keep() when (@Keep = Keep12PX) {}
}

.Keep(@Keep) when (lightness(@Keep) >= 12px) and (@Keep > 0) {}
.Keep(@Keep) when (lightness(@Keep) >= Keep12PX) and (@Keep > @Keep12E5) {}

.Keep(@Keep: 12px; @Keep: @Keep12PX; ...) /*:*/ {}
.Keep(@Keep: "12PX"; @Keep: "12PX"; ...) /*:*/ {}

@media (min-width: 700px) /*:*/ {}

@keyframes identifier {
  from {
    margin-top: 50px;
  }
  50% {
    margin-top: 150px;
  }
  to {
    margin-top: 100px;
  }
}

@-webkit-keyframes float-up {
  from {
    margin-top: 50px;
  }
  50% {
    margin-top: 150px;
  }
  to {
    margin-top: 100px;
  }
}

.foo {
  color: hsl(0.75turn, 60%, 70%);
}

p:first-child {
  color: lime;
  background-color: black;
  padding: 5px;
}

a::after {
  content: "→";
}

a:after {
  content: "→";
}

::-webkit-progress-bar {
  background-color: orange;
}

table {}

.foo {
  &-KeepSelector {}
  &-KeepSelector & .KeepClassSelector {}
  &-100\.200 {}
}

